library/Freckle/App/OpenTelemetry.hs

```diff
@@ -1,18 +1,67 @@
+-- | Application tracing via <https://opentelemetry.io/>
+--
+-- @
+-- data App = App
+--   { -- ...
+--   , appTracer :: Tracer
+--   }
+--
+-- instance HasTracer App where
+--   tracerL = lens appTracer $ \x y -> x { appTracer = y }
+--
+-- loadApp f = do
+--   -- ...
+--   withTracerProvider $ \tracerProvider -> do
+--     let appTracer = makeTracer tracerProvider "my-app" tracerOptions
+--     f App {..}
+-- @
+--
+-- You may need to do this even if you don't plan to manually trace things, in order to
+-- satisfy the 'MonadTracer' constraint required by functions like 'runDB'. If
+-- you don't need this feature, and don't plan on running an otel-collector, set
+-- @OTEL_TRACES_EXPORTER=none@ in the environment, which makes all tracing a
+-- no-op.
+--
+-- In the future, it should be possible to use @OTEL_SDK_DISABLED@ for the same
+-- purpose. See <https://github.com/iand675/hs-opentelemetry/issues/60>.
 module Freckle.App.OpenTelemetry
-  ( MonadTracer (..)
+  ( HasTracer (..)
+  , Tracer
+
+    -- * Effects
+  , MonadTracer (..)
+  , inSpan
+  , defaultSpanArguments
+
+    -- * Querying
+  , getCurrentTraceId
+
+    -- * Setup
+  , withTracerProvider
+
+    -- ** 'Tracer'
+  , makeTracer
+  , tracerOptions
   ) where
 
 import Freckle.App.Prelude
 
-import Network.AWS.XRayClient.WAI (XRayVaultData, vaultDataFromRequest)
-import Yesod.Core (HandlerFor, waiRequest)
+import OpenTelemetry.Context (lookupSpan)
+import OpenTelemetry.Context.ThreadLocal (getContext)
+import OpenTelemetry.Trace hiding (inSpan)
+import OpenTelemetry.Trace.Core (getSpanContext)
+import qualified OpenTelemetry.Trace.Core as Trace (SpanContext (..))
+import OpenTelemetry.Trace.Id (SpanId, TraceId)
+import OpenTelemetry.Trace.Monad
+import UnliftIO.Exception (bracket)
 
--- | Class for reading 'XRayVaultData'
---
--- This is named the same as the OpenTelemetry class we'll use once we move to
--- that tracing system
-class MonadTracer m where
-  getVaultData :: m (Maybe XRayVaultData)
+withTracerProvider :: MonadUnliftIO m => (TracerProvider -> m a) -> m a
+withTracerProvider =
+  bracket
+    (liftIO initializeGlobalTracerProvider)
+    (liftIO . shutdownTracerProvider)
 
-instance MonadTracer (HandlerFor app) where
-  getVaultData = vaultDataFromRequest <$> waiRequest
+getCurrentTraceId :: MonadIO m => m (Maybe (TraceId, SpanId))
```

Comment:
Minor naming quibble: I you don't want to go so wordy as `getCurrentTraceAndSpanId`, why not `getCurrentSpanId` since that's the more specific of the two pieces of information here?

Reply:
I now forget why this returns both. It's only used for the trace-id part anyway, which is probably where the naming came from. Let me clean this up a bit.